File: Backend/src/handlers/modelHandler.js
'use strict';

const Path = require('path');
const Fs = require('fs');
const tf = require('@tensorflow/tfjs-node');
const config = require('../config');
const logger = require('../utils/logger');

// Global variables for loaded models
let landmarkModel = null;
let videoLstmModel = null;
let videoTransformerModel = null;
let imageClassMapping = {};
let videoClassMapping = {};

/**
 * Load all models and class mappings
 * @returns {Promise<Object>} Status of loaded models
 */
async function loadModels() {
  try {
    logger.info('Loading models...');
    
    // Load class mappings
    const imageClassMappingPath = Path.join(
      config.models.directory, 
      config.models.files.imageClassMapping
    );
    
    const videoClassMappingPath = Path.join(
      config.models.directory, 
      config.models.files.videoClassMapping
    );
    
    if (Fs.existsSync(imageClassMappingPath)) {
      imageClassMapping = JSON.parse(Fs.readFileSync(imageClassMappingPath, 'utf8'));
      logger.info('Image class mapping loaded.');
    } else {
      logger.warn('Image class mapping file not found:', imageClassMappingPath);
    }
    
    if (Fs.existsSync(videoClassMappingPath)) {
      videoClassMapping = JSON.parse(Fs.readFileSync(videoClassMappingPath, 'utf8'));
      logger.info('Video class mapping loaded.');
    } else {
      logger.warn('Video class mapping file not found:', videoClassMappingPath);
    }
    
    // Load models
    const landmarkModelPath = Path.join(
      config.models.directory, 
      config.models.files.landmarkModel
    );
    
    const videoLstmModelPath = Path.join(
      config.models.directory, 
      config.models.files.videoLstmModel
    );
    
    const videoTransformerModelPath = Path.join(
      config.models.directory, 
      config.models.files.videoTransformerModel
    );
    
    if (Fs.existsSync(landmarkModelPath)) {
      try {
        landmarkModel = await tf.loadLayersModel(`file://${landmarkModelPath}`);
        logger.info('Landmark model loaded.');
      } catch (error) {
        logger.error('Error loading landmark model:', error);
      }
    } else {
      logger.warn('Landmark model not found:', landmarkModelPath);
    }
    
    if (Fs.existsSync(videoLstmModelPath)) {
      try {
        videoLstmModel = await tf.loadLayersModel(`file://${videoLstmModelPath}`);
        logger.info('Video LSTM model loaded.');
      } catch (error) {
        logger.error('Error loading video LSTM model:', error);
      }
    } else {
      logger.warn('Video LSTM model not found:', videoLstmModelPath);
    }
    
    if (Fs.existsSync(videoTransformerModelPath)) {
      try {
        videoTransformerModel = await tf.loadLayersModel(`file://${videoTransformerModelPath}`);
        logger.info('Video transformer model loaded.');
      } catch (error) {
        logger.error('Error loading video transformer model:', error);
      }
    } else {
      logger.warn('Video transformer model not found:', videoTransformerModelPath);
    }
    
    logger.info('Models loaded successfully.');
    
    return {
      landmarkModel: landmarkModel !== null,
      videoLstmModel: videoLstmModel !== null,
      videoTransformerModel: videoTransformerModel !== null,
      imageClassMapping: Object.keys(imageClassMapping).length > 0,
      videoClassMapping: Object.keys(videoClassMapping).length > 0
    };
  } catch (error) {
    logger.error('Error loading models:', error);
    throw error;
  }
}

/**
 * Predict static sign (letter) from landmarks
 * @param {Array} landmarks - Flattened array of hand landmarks
 * @returns {Promise<Object>} Prediction result
 */
async function predictStaticSign(landmarks) {
  if (!landmarkModel) {
    throw new Error('Landmark model not loaded');
  }
  
  // Convert landmarks to tensor
  const tensor = tf.tensor2d([landmarks], [1, config.models.params.numLandmarkFeatures]);
  
  // Predict
  const predictions = await landmarkModel.predict(tensor).data();
  
  // Get the highest prediction
  let maxIndex = 0;
  let maxConfidence = 0;
  
  for (let i = 0; i < predictions.length; i++) {
    if (predictions[i] > maxConfidence) {
      maxConfidence = predictions[i];
      maxIndex = i;
    }
  }
  
  // Get the class name
  const predictedClass = imageClassMapping[maxIndex] || 'Unknown';
  
  // Clean up
  tensor.dispose();
  
  return {
    class: predictedClass,
    confidence: maxConfidence,
    index: maxIndex
  };
}

/**
 * Predict dynamic sign (word) from landmark sequence
 * @param {Array} landmarkSequence - Array of landmark arrays
 * @param {string} modelChoice - 'lstm' or 'transformer'
 * @returns {Promise<Object>} Prediction result
 */
async function predictDynamicSign(landmarkSequence, modelChoice = 'transformer') {
  // Choose model based on modelChoice
  const model = modelChoice === 'transformer' ? videoTransformerModel : videoLstmModel;
  
  if (!model) {
    throw new Error(`${modelChoice} model not loaded`);
  }
  
  // Ensure sequence has correct dimensions
  let processedSequence = landmarkSequence;
  const numFrames = config.models.params.numFramesVideo;
  const numFeatures = config.models.params.numLandmarkFeatures;
  
  // Pad or truncate sequence to correct length
  if (landmarkSequence.length < numFrames) {
    // Pad with zeros
    processedSequence = [
      ...landmarkSequence,
      ...Array(numFrames - landmarkSequence.length).fill(Array(numFeatures).fill(0))
    ];
  } else if (landmarkSequence.length > numFrames) {
    // Truncate
    processedSequence = landmarkSequence.slice(0, numFrames);
  }
  
  // Convert to tensor
  const tensor = tf.tensor3d([processedSequence], [1, numFrames, numFeatures]);
  
  // Predict
  const predictions = await model.predict(tensor).data();
  
  // Get the highest prediction
  let maxIndex = 0;
  let maxConfidence = 0;
  
  for (let i = 0; i < predictions.length; i++) {
    if (predictions[i] > maxConfidence) {
      maxConfidence = predictions[i];
      maxIndex = i;
    }
  }
  
  // Get the class name
  const predictedClass = videoClassMapping[maxIndex] || 'Unknown';
  
  // Clean up
  tensor.dispose();
  
  return {
    class: predictedClass,
    confidence: maxConfidence,
    index: maxIndex,
    modelUsed: modelChoice
  };
}

/**
 * Convert text to sign language
 * @param {string} text - Input text
 * @returns {Object} Sign language representation
 */
function textToSign(text) {
  if (!text) {
    throw new Error('No text provided');
  }
  
  // Parse text and convert to signs
  const words = text.trim().toLowerCase().split(/\s+/);
  const result = [];
  
  for (const word of words) {
    // Check if word exists in video class mapping (known words)
    let isKnownWord = false;
    let mappedWord = word;
    
    for (const [key, value] of Object.entries(videoClassMapping)) {
      if (value.toLowerCase() === word) {
        isKnownWord = true;
        mappedWord = value; // Use the exact case from the mapping
        break;
      }
    }
    
    if (isKnownWord) {
      // Known word - can be represented as a single sign
      result.push({
        type: 'word',
        original: word,
        mapped: mappedWord,
        knownInDataset: true
      });
    } else {
      // Unknown word - needs to be fingerspelled
      const letters = [];
      
      for (const letter of word) {
        // Check if we have this letter in our image classes
        let letterExists = false;
        let mappedLetter = letter.toUpperCase();
        
        // Look through the image class mapping for this letter
        for (const [key, value] of Object.entries(imageClassMapping)) {
          if (value.toLowerCase() === letter.toLowerCase()) {
            letterExists = true;
            mappedLetter = value; // Use the exact case from the mapping
            break;
          }
        }
        
        letters.push({
          letter: letter,
          mapped: mappedLetter,
          exists: letterExists
        });
      }
      
      result.push({
        type: 'fingerspell',
        original: word,
        letters: letters
      });
    }
  }
  
  return {
    text: text,
    signs: result
  };
}

/**
 * Get available words from the vocabulary
 * @returns {Array} List of available words
 */
function getAvailableWords() {
  return Object.entries(videoClassMapping).map(([key, value]) => ({
    id: key,
    word: value
  }));
}

/**
 * Get available letters from the vocabulary
 * @returns {Array} List of available letters
 */
function getAvailableLetters() {
  return Object.entries(imageClassMapping).map(([key, value]) => ({
    id: key,
    letter: value
  }));
}

/**
 * Get model status
 * @returns {Object} Status of loaded models
 */
function getModelStatus() {
  return {
    landmarkModel: landmarkModel !== null,
    videoLstmModel: videoLstmModel !== null,
    videoTransformerModel: videoTransformerModel !== null,
    imageClassMapping: Object.keys(imageClassMapping).length > 0,
    videoClassMapping: Object.keys(videoClassMapping).length > 0
  };
}

module.exports = {
  loadModels,
  predictStaticSign,
  predictDynamicSign,
  textToSign,
  getAvailableWords,
  getAvailableLetters,
  getModelStatus
};

File: Backend/handlers/model_handler.py
import os
import json
import numpy as np
import tensorflow as tf
from typing import List, Dict, Union, Any, Optional
from fastapi import HTTPException
from config import Config
import traceback # 

def log_info(message: str, meta: Optional[Dict] = None):
    print(f"[INFO] {message}", meta if meta else "")

def log_error(message: str, error: Union[Exception, Dict] = None):
    meta = {"message": str(error), "stack": traceback.format_exc()} if isinstance(error, Exception) else error
    print(f"[ERROR] {message}", meta if meta else "")

class ModelHandler:
    def __init__(self):
        self.static_sign_model = None
        self.video_lstm_model = None
        self.image_class_mapping = {}
        self.video_class_mapping = {}
        self.models_loaded = False

    async def load_models(self):
        log_info('Loading TensorFlow models and class mappings...')
        try:
            def _load_model_h5_only(h5_path, model_name, custom_objects=None):
                model = None
                if os.path.exists(h5_path):
                    log_info(f"Attempting to load {model_name} from H5: {h5_path}")
                    try:
                        model = tf.keras.models.load_model(h5_path, custom_objects=custom_objects)
                        log_info(f"{model_name} loaded successfully from H5.")
                        return model
                    except Exception as e:
                        log_error(f"Failed to load {model_name} from H5 at {h5_path}. Error: {e}.", e)
                        raise FileNotFoundError(f"Failed to load {model_name} from H5: {e}")
                else:
                    raise FileNotFoundError(f"H5 model file not found for {model_name} at {h5_path}.")
                
                return model 

            self.static_sign_model = _load_model_h5_only(
                Config.H5_LANDMARK_MODEL_PATH,
                "Static Sign Model (Landmark Model)"
            )

            self.video_lstm_model = _load_model_h5_only(
                Config.H5_VIDEO_LSTM_MODEL_PATH,
                "Video LSTM Model"
            )

            with open(Config.IMAGE_CLASS_MAPPING_PATH, 'r') as f:
                self.image_class_mapping = {int(k): v for k, v in json.load(f).items()}
            log_info('Image Class Mapping loaded successfully.')

            with open(Config.VIDEO_CLASS_MAPPING_PATH, 'r') as f:
                self.video_class_mapping = {int(k): v for k, v in json.load(f).items()}
            log_info('Video Class Mapping loaded successfully.')
            
            self.models_loaded = True
            return {
                "landmark_model": bool(self.static_sign_model),
                "video_lstm_model": bool(self.video_lstm_model),
                "image_class_mapping": bool(self.image_class_mapping),
                "video_class_mapping": bool(self.video_class_mapping)
            }
        except Exception as e:
            log_error('Error loading models or class mappings:', e)
            self.models_loaded = False
            raise

    def get_model_status(self) -> Dict[str, bool]:
        return {
            "landmark_model": self.static_sign_model is not None,
            "video_lstm_model": self.video_lstm_model is not None,
            "video_transformer_model": False, # Always False as it's removed
            "image_class_mapping": bool(self.image_class_mapping),
            "video_class_mapping": bool(self.video_class_mapping)
        }

    async def predict_static_sign(self, landmarks: List[float]) -> Dict[str, Union[str, float, int, None]]:
        if not self.static_sign_model:
            log_error("Static sign model not loaded for prediction.")
            raise HTTPException(status_code=500, detail="Static sign model not loaded.")
        if not landmarks or len(landmarks) != Config.NUM_LANDMARK_FEATURES:
            raise HTTPException(status_code=400, detail=f"Invalid landmark array length. Expected {Config.NUM_LANDMARK_FEATURES}, got {len(landmarks)}.")

        try:
            input_tensor = tf.constant([landmarks], dtype=tf.float32)
            predictions = self.static_sign_model.predict(input_tensor, verbose=0)
            probabilities = tf.nn.softmax(predictions).numpy()[0]

            predicted_index = np.argmax(probabilities)
            predicted_class = self.image_class_mapping.get(predicted_index, "Unknown")
            confidence = float(probabilities[predicted_index])

            log_info(f"Static sign prediction: Class={predicted_class}, Confidence={confidence:.4f}")

            return {
                "class": predicted_class,
                "confidence": confidence,
                "index": int(predicted_index)
            }
        except Exception as e:
            log_error("Error in predict_static_sign:", e)
            raise HTTPException(status_code=500, detail=f"Prediction failed: {e}")

    async def predict_dynamic_sign(self, landmark_sequence: List[List[float]]) -> Dict[str, Union[str, float, int, None]]:
        model_to_use = self.video_lstm_model

        if not model_to_use:
            log_error(f"Video LSTM model not loaded for prediction.")
            raise HTTPException(status_code=500, detail=f"Video LSTM model not loaded.")

        if not landmark_sequence or len(landmark_sequence) == 0:
            raise HTTPException(status_code=400, detail='Landmark sequence is empty.')

        num_frames_expected = Config.NUM_FRAMES_VIDEO
        num_features = Config.NUM_LANDMARK_FEATURES

        processed_sequence = list(landmark_sequence)

        if len(processed_sequence) > num_frames_expected:
            processed_sequence = processed_sequence[len(processed_sequence) - num_frames_expected:]
        elif len(processed_sequence) < num_frames_expected:
            padding = [0.0] * num_features
            while len(processed_sequence) < num_frames_expected:
                processed_sequence.append(padding)
        
        processed_sequence = [
            (frame[:num_features] + [0.0] * (num_features - len(frame))) if len(frame) != num_features
            else frame
            for frame in processed_sequence
        ]
        
        try:
            input_tensor = tf.constant([processed_sequence], dtype=tf.float32)
            predictions = model_to_use.predict(input_tensor, verbose=0)
            probabilities = tf.nn.softmax(predictions).numpy()[0]

            predicted_index = np.argmax(probabilities)
            predicted_class = self.video_class_mapping.get(predicted_index, "Unknown")
            confidence = float(probabilities[predicted_index])

            log_info(f"Dynamic sign prediction (LSTM): Class={predicted_class}, Confidence={confidence:.4f}")

            return {
                "class": predicted_class,
                "confidence": confidence,
                "index": int(predicted_index),
                "modelUsed": "lstm"
            }
        except Exception as e:
            log_error("Error in predict_dynamic_sign:", e)
            raise HTTPException(status_code=500, detail=f"Prediction failed: {e}")

    def text_to_sign(self, text: str) -> Dict[str, Any]:
        if not text:
            raise HTTPException(status_code=400, detail="No text provided")
        
        words = text.strip().lower().split()
        result_signs = []

        for word in words:
            is_known_word = False
            mapped_word = word
            
            for k, v in self.video_class_mapping.items():
                if v.lower() == word:
                    is_known_word = True
                    mapped_word = v
                    break
            
            if is_known_word:
                result_signs.append({
                    "type": "word",
                    "original": word,
                    "mapped": mapped_word,
                    "knownInDataset": True
                })
            else:
                letters = []
                for letter in word:
                    letter_exists = False
                    mapped_letter = letter.upper()
                    
                    for k, v in self.image_class_mapping.items():
                        if v.lower() == letter.lower():
                            letter_exists = True
                            mapped_letter = v
                            break
                    
                    letters.append({
                        "letter": letter,
                        "mapped": mapped_letter,
                        "exists": letter_exists
                    })
                
                result_signs.append({
                    "type": "fingerspell",
                    "original": word,
                    "letters": letters
                })
        
        return {
            "text": text,
            "signs": result_signs
        }

    def get_available_words(self) -> List[Dict[str, Union[int, str]]]:
        return [{"id": k, "word": v} for k, v in self.video_class_mapping.items()]

    def get_available_letters(self) -> List[Dict[str, Union[int, str]]]:
        return [{"id": k, "letter": v} for k, v in self.image_class_mapping.items()]